File: program_stack/algorithms/sorting/bubble_sort/cpp/bubble_sort.cpp
#include <algorithm>
#include <cstdlib>
#include <iostream>
#include <vector>

// In-place bubble sort (non-decreasing). Stable.
void sort(std::vector<int>& a) {
    if (a.size() < 2) {
        return;
    }
    for (std::size_t end = a.size() - 1; end > 0; --end) {
        bool swapped = false;
        for (std::size_t i = 0; i < end; ++i) {
            if (a[i] > a[i + 1]) {
                std::swap(a[i], a[i + 1]);
                swapped = true;
            }
        }
        if (!swapped) {
            break;
        }
    }
}

static void expect(std::vector<int> input, const std::vector<int>& want) {
    sort(input);
    if (input != want) {
        std::cerr << "mismatch\n";
        std::exit(1);
    }
}

int main() {
    expect({}, {});
    expect({42}, {42});
    expect({1, 2, 3}, {1, 2, 3});
    expect({3, 2, 1}, {1, 2, 3});
    expect({5, 1, 4, 2, 8}, {1, 2, 4, 5, 8});
    expect({3, 1, 2, 1, 3}, {1, 1, 2, 3, 3});
    std::cout << "bubble_sort: ok\n";
    return 0;
}

File: program_stack/algorithms/sorting/bubble_sort/java/BubbleSort.java
import java.util.Arrays;

public class BubbleSort {
    /** In-place bubble sort (non-decreasing). Stable. */
    public static void sort(int[] a) {
        if (a == null || a.length < 2) {
            return;
        }
        int n = a.length;
        for (int end = n - 1; end > 0; end--) {
            boolean swapped = false;
            for (int i = 0; i < end; i++) {
                if (a[i] > a[i + 1]) {
                    int tmp = a[i];
                    a[i] = a[i + 1];
                    a[i + 1] = tmp;
                    swapped = true;
                }
            }
            if (!swapped) {
                break;
            }
        }
    }

    private static void expect(int[] input, int[] want) {
        int[] a = Arrays.copyOf(input, input.length);
        sort(a);
        if (!Arrays.equals(a, want)) {
            System.err.println("got " + Arrays.toString(a) + " want " + Arrays.toString(want));
            System.exit(1);
        }
    }

    public static void main(String[] args) {
        expect(new int[] {}, new int[] {});
        expect(new int[] {42}, new int[] {42});
        expect(new int[] {1, 2, 3}, new int[] {1, 2, 3});
        expect(new int[] {3, 2, 1}, new int[] {1, 2, 3});
        expect(new int[] {5, 1, 4, 2, 8}, new int[] {1, 2, 4, 5, 8});
        expect(new int[] {3, 1, 2, 1, 3}, new int[] {1, 1, 2, 3, 3});
        System.out.println("bubble_sort: ok");
    }
}
